internal/webhook/external/azure.go

```diff
@@ -0,0 +1,181 @@
+// getSecretValues implements WebhookReceiver.
+func (q *azureWebhookReceiver) getSecretValues(
+	secretData map[string][]byte,
+) ([]string, error) {
+	secretValue, ok := secretData[azureSecretDataKey]
+	if !ok {
+		return nil,
+			errors.New("Secret data is not valid for an Azure WebhookReceiver")
+	}
+	return []string{string(secretValue)}, nil
+}
+
+// getHandler implements WebhookReceiver.
+func (q *azureWebhookReceiver) getHandler(requestBody []byte) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := logging.LoggerFromContext(ctx)
+		logger.Debug("identifying source repository")
+
+		logger.Info("received Azure webhook event",
+			"requestBody", string(requestBody),
+		)
+
+		var event azureEvent
+		if err := json.Unmarshal(requestBody, &event); err != nil {
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(errors.New("invalid request body"),
+					http.StatusBadRequest,
+				),
+			)
+			return
+		}
+
+		switch {
+		case event.Action == "ping":
+			xhttp.WriteResponseJSON(
+				w,
+				http.StatusOK,
+				map[string]string{
+					"msg": "ping event received, webhook is configured correctly",
+				},
+			)
+			return
+		case event.Action == "push":
+			repoURL := fmt.Sprintf("%s/%s",
+				event.Request.Host,
+				event.Target.Repository,
+			)
+
+			// Payloads from azure contain no information about media type, so we
+			// normalize the URL BOTH as if it were an image repo URL and as if it were
+			// a chart repository URL. These will coincidentally be the same, but by
+			// doing this, we safeguard against future changes to normalization logic.
+			// Note: The refresh logic will dedupe the URLs, so this does not create
+			// the possibility of a double refresh.
+			repoURLs := []string{
+				image.NormalizeURL(repoURL),
+				helm.NormalizeChartRepositoryURL(repoURL),
+			}
+
+			logger = logger.WithValues("repoURLs", repoURLs)
+			ctx = logging.ContextWithLogger(ctx, logger)
+
+			refreshWarehouses(ctx, w, q.client, q.project, repoURL)
+			// For Azure DevOps, when you configure an action trigger
+			// and test it through their UI, it sends an event type
+			// that is the same as the one you configured it for
+			// as opposed to a "ping" event like Azure Container Registry does.
+		case event.EventType == "git.push":
+			repoURL := git.NormalizeURL(event.Resource.Repository.RemoteURL)
+			logger = logger.WithValues("repoURL", repoURL)
+			ctx = logging.ContextWithLogger(ctx, logger)
+			refreshWarehouses(ctx, w, q.client, q.project, repoURL)
+		default:
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(
+					fmt.Errorf("event type %s is not supported", event.Action),
+					http.StatusNotImplemented,
+				),
+			)
+			return
+		}
+	})
+}
+
+// events can come from Azure Container Registry or Azure DevOps.
+// azureEvent is a struct that represents the payload of an event from either
+// of these services as there are no overlapping fields between the two
```

Comment:
By widely observed Go convention, comments on functions should start with the function name:

```suggestion
// azureEvent represents the payload of an event from either Azure Container
// Registry or Azure DevOps.
// of these services as there are no overlapping fields between the two
```